doc/sphinx/src/using-closures.rst

```diff
@@ -3,131 +3,362 @@
-finite volume cell may have more than one material in it. In this
-situation, one must decide how to compute thermodynamic quantities in
-that cell. One choice, though not necessarily the optimal choice, is
-to assume pressure-temperature equilibrium (PTE). This implies that there's
-a single pressure in the cell and a single temperature, which are the
-same for all materials.
-
-``singularity-eos`` provides several methods for finding a PTE
-soluution. These methods differ in what they treat as independent
-variables, and thus what precise system of equations they solve
-for. However they all share some common characteristics.
+In the single-material Euler equations, the mass and energy are typically
+evolved and the EOS is called to provide a pressure for the momentum equations.
+When transitioning to a multi-material approach, a single velocity is typically
+used and the Euler equations are solved with respect to the bulk fluid motion.
+In this case, the pressure contribution to momentum isn't well-defined and in
+principle each material could have its own pressure contribution to material
+motion. Furthermore, the paritioning of volume and energy between the materials
+in the flow is not well-defined either.
+
+As a result, a multi-material closure rule is needed to determine both how to
+compute the pressure response of the flow and how to partition the volume and
+energy between the individual materials. In this situation, one must decide how
+to compute thermodynamic quantities in that cell for each material.
 
 Governing Equations and Assumptions
 ------------------------------------
 
-Given a set of :math:`N` materials ranging from :math:`i = 0` to
-:math:`N-1`, each material can, in principle, have a distinct pressure
-:math:`P_i`, a distinct temperature :math:`T_i`, a distinct material
-density :math:`\rho_i`, and a distinct specific material energy
-:math:`\varepsilon_i`. For some finite total volume :math:`V`, each
-material also occupies some fraction of that volume given by the
-*volume fraction* :math:`f_i` such that
+In a general sense then the mixed
+material closure rule takes the form
 
 .. math::
 
-  \sum_{i=0}^{N - 1} f_i = 1
+  P_i = F(\rho, \epsilon, \mu_1, ..., \mu_i, ..., \mu_{N-1}) \\
+  \rho_i = G(\rho, \epsilon, \mu_1, ..., \mu_i, ..., \mu_{N-1}) \\
+  \epsilon_i = H(\rho, \epsilon, \mu_1, ..., \mu_i, ..., \mu_{N-1}) \\
+  T_i = J(\rho, \epsilon, \mu_1, ..., \mu_i, ..., \mu_{N-1})
 
-The average density in a given material-occupied volume is
+where each material has its own density, :math:`\rho_i`, and specific internal
+energy, :math:`e_i`, as well as in principle its own pressure and temperature.
+Each will be a function of the bulk density, :math:`\rho`, the bulk specific
+internal energy, :math:`\epsilon`, and both the mass fraction, :math:`\mu_j`,
+of all materials present.
+
+For convenience of minimizing input arguments though, the solvers in
```

Comment:
Divide it up in \bar{rho} and a guess for f_i would be better I think. Are we really sure we want to stay with these two coupled variables for the future? It is much easier to rewrite the code now than later when we have connected a bunch of host codes that uses this.

Reply:
No I don't think we're bound to this input specification at all, and I think specifying `rho_bar` directly is an excellent idea since it clearly differentiates inputs from outputs.

I agree that it will be easier to change things sooner rather than later, but since we already have codes that depend on the current input specification, I'd imagine that we'd offer both the existing specification as well as a newer, less confusing one in the future. In fact, I'd love to rewrite the `get_sg_eos()` functions to use what you suggest in the future.

Right now I don't really have the bandwidth to tackle this though, so I'm afraid I'll have to punt on code changes here. I'll create an issue though:
https://github.com/lanl/singularity-eos/issues/360

Reply:
I want to think about this a little bit more before we jump on this change. I agree this suggestion is less confusing, but we should think about exactly the optimal set of inputs/outputs for a host code.

Reply:
So, when it comes to KPT the best would be using the mass fractions since that is what gets updated with time. 
For computational reasons we should stay at what gives the best derivatives. Nothing will improve convergence better than using the best possible derivatives.

Reply:
> So, when it comes to KPT the best would be using the mass fractions since that is what gets updated with time.

This makes sense---that also motivates using rhobar over rho, since that's what Eulerian codes update with time.

> For computational reasons we should stay at what gives the best derivatives. Nothing will improve convergence better than using the best possible derivatives.

This is a discussion for another time but... while this seems intuitively right, but is that always true? For example, for tables that combine multiple phases together, are the "best" derivatives not discontinuous at phase transitions almost by definition?

Reply:
Just to be clear, I'm not proposing changing the existing inputs. I'm proposing modifying the classes to accept different multiple input specifications.